ath79: add support for Mikrotik LHG 5

XDjackieXD opened the pull request:
This commit adds support for the MikroTik LHG 5 series of devices which were already supported in ar71xx.
The patch was tested on a RBLHG-5HPnD-XL and is working as intended.
Currently the TX power is limited to 20dBm even on the HP devices as there is no differentiation between the HP and non-HP devices.



The MikroTik LHG 5 series (product codes RBLHG-5nD, RBLHG-5HPnD and
RBLHG-5HPnD-XL) devices are an outdoor 5GHz CPE with a 24.5dBi or 27dBi
integrated antenna built around the Atheros AR9344 SoC.
It is very similar to the SXT Lite5 series which this patch is based
upon.

Specifications:
 - SoC: Atheros AR9344
 - RAM: 64 MB
 - Storage: 16 MB SPI NOR
 - Wireless: Atheros AR9340 (SoC) 802.11a/n 2x2:2
 - Ethernet: Atheros AR8229 switch (SoC), 1x 10/100 port,
    8-32 Vdc PoE in
 - 8 user-controllable LEDs:
  - 1x power (blue)
  - 1x user (white)
  - 1x ethernet (green)
  - 5x rssi (green)

 See https://mikrotik.com/product/RBLHG-5nD for more details.

Notes:
 The device was already supported in the ar71xx target.

Flashing:
 TFTP boot initramfs image and then perform a sysupgrade. Follow common
 MikroTik procedure as in https://openwrt.org/toh/mikrotik/common.

Signed-off-by: Jakob (Jack/XDjackieXD) <jakob@chaosfield.at>

rogerpueyo reviewed (commented):
Dear @XDjackieXD ,

Thanks for your PR. The LHG 5 seems to be similar to the LHG 2nD, which is already supported, so you may want to do it in a similar way.

For example, the [DTS file for the LHG 2nD](https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9533_mikrotik_routerboard-lhg-2nd.dts) is based on the [common DTSI file for the LHG-HB platform](https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9533_mikrotik_routerboard-lhg-hb.dtsi); you may want to go the same way.

I spotted some strange things in your DTS. For example, you declared this:
```
		power {
			label = "green:power";
			gpios = <&gpio 11 GPIO_ACTIVE_HIGH>;
			default-state = "on";
		};
[...]
		led_user: user {
			label = "green:user";
			gpios = <&gpio 20 GPIO_ACTIVE_LOW>;
		};
```
but it doesn't match the commit message, which says:
```
  - 1x power (blue)
  - 1x user (white)
```
This needs to be coherent.

Cheers

XDjackieXD commented:
> Thanks for your PR. The LHG 5 seems to be similar to the LHG 2nD, which is already supported, so you may want to do it in a similar way.
> 
> For example, the [DTS file for the LHG 2nD](https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9533_mikrotik_routerboard-lhg-2nd.dts) is based on the [common DTSI file for the LHG-HB platform](https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9533_mikrotik_routerboard-lhg-hb.dtsi); you may want to go the same way.

I thought about this but as the 2.4GHz and 5GHz Versions of the LHG Series uses a different SoC I'm not sure if this is apropriate (in theory I could include the lhg-hb dtsi and overwrite the compatible= with the correct SoC but that could be confusing because I'm then including a dts file which has a filename that is prefixed with another SoC name).
Maybe rename `qca9533_mikrotik_routerboard-16m.dtsi` to be SoC independent and include this in my dts?


> I spotted some strange things in your DTS. For example, you declared this:
> 
> ```
> 		power {
> 			label = "green:power";
> 			gpios = <&gpio 11 GPIO_ACTIVE_HIGH>;
> 			default-state = "on";
> 		};
> [...]
> 		led_user: user {
> 			label = "green:user";
> 			gpios = <&gpio 20 GPIO_ACTIVE_LOW>;
> 		};
> ```
> 
> but it doesn't match the commit message, which says:
> 
> ```
>   - 1x power (blue)
>   - 1x user (white)
> ```
> 
> This needs to be coherent.

Ah thanks. I missed that when changing the copied dts from the SXTsq. Will fix that with the next push.

XDjackieXD commented:
I split the parts that most likely affect more devices into a `ar9344_mikrotik_routerboard-16m-nor.dtsi` file. The sxtSq which currently is the only other mikrotik device with this chip in ath79 does not use it though as it uses a NAND flash for rootfs storage and therefore uses a different layout for the SPI NOR flash.

adschm commented on target/linux/ath79/dts/ar9344_mikrotik_routerboard-lhg-5nd.dts:
```diff
@@ -0,0 +1,92 @@
+
+	compatible = "syscon", "simple-mfd";
+};
+
```
Remove empty line at EOF.

hauke commented:
Thank you for your patch, I applied it to master in 48774decea9b42e97edd53ce33e574de7d85f3df.

hauke commented:
I reverted the commit again in cee0a95005a1a3ef48f5c2750605140dae6b7e66 Please update the commit with a valid name in the Signed-off-by line.

adschm commented on target/linux/ath79/dts/ar9344_mikrotik_routerboard-lhg-5nd.dts:
```diff
@@ -0,0 +1,91 @@
+
+/ {
+	compatible = "mikrotik,routerboard-lhg-5nd", "qca,ar9344";
+	model = "MikroTik LHG 5nD platform";
```
Since this has been reopened and needs to be updated anyway, I wonder why the model here is different from DEVICE_MODEL below?

XDjackieXD replied:
Mainly because I followed the pattern I saw wit hthe other mikrotik devices for both (to be specific: the `DEVICE_MODEL` string comes from the LHG2 and the `model` string is based on the string in `ar9344_mikrotik_routerboard-sxt-5n.dtsi` which the dts file for the LHG5 is mainly based on).
Do you have suggestion what string to use to unify the two?

rogerpueyo replied:
I'm sharing this here, just in case it can help in some way.

MikroTik often sells two different products with very similar hardware. Sometimes they add an extra Ethernet port, sometimes they only differ in the antenna gain, etc. I recently spotted this with the ipq40xx-based "SXTsq 5ac" and the "Disc Lite5 ac". In the case of this devices, under /sys/firmware/mikrotik/hard_config, you can see that the "board_identifier" is the same, whereas differences appear in "board_product_code" and "product_name", and sometimes in "hw_options".

My idea with the "platform" thing was to use this word only in the common .dtsi file shared by different devices with the same "board_identifier", for purely informative purpose inside the .dtsi file itself. Then, the actual .dts file for each specific device would override the model name.

I suggest to add support for the LHG 5 using the device name, on a single .dts file. Eventually, if a second device appears using the same "board_identifier", a common .dtsi can be created, and then use each .dts to just set each device-specific parameters.

Cheers!

XDjackieXD replied:
I set it now to be the same as DEVICE_MODEL.
There is a dtsi file where I tried to split out all the options that similar devices will probably share.

XDjackieXD committed:
ath79: add support for Mikrotik LHG 5

The MikroTik LHG 5 series (product codes RBLHG-5nD, RBLHG-5HPnD and
RBLHG-5HPnD-XL) devices are an outdoor 5GHz CPE with a 24.5dBi or 27dBi
integrated antenna built around the Atheros AR9344 SoC.
It is very similar to the SXT Lite5 series which this patch is based
upon.

Specifications:
 - SoC: Atheros AR9344
 - RAM: 64 MB
 - Storage: 16 MB SPI NOR
 - Wireless: Atheros AR9340 (SoC) 802.11a/n 2x2:2
 - Ethernet: Atheros AR8229 switch (SoC), 1x 10/100 port,
    8-32 Vdc PoE in
 - 8 user-controllable LEDs:
  - 1x power (blue)
  - 1x user (white)
  - 1x ethernet (green)
  - 5x rssi (green)

 See https://mikrotik.com/product/RBLHG-5nD for more details.

Notes:
 The device was already supported in the ar71xx target.

Flashing:
 TFTP boot initramfs image and then perform a sysupgrade. Follow common
 MikroTik procedure as in https://openwrt.org/toh/mikrotik/common.

Signed-off-by: Jakob Riepler <jakob+openwrt@chaosfield.at>

XDjackieXD commented:
Updated with correct signed-off line and corected device model string in the dts file.

hauke commented:
Thank you for your patch, I applied it to master in e0683839b8036388213d7662f3e0066a29b7d480.
